app/assets/javascripts/beak/wip-listener.coffee

```diff
@@ -0,0 +1,142 @@
+import { DiskSource, NewSource } from  './nlogo-source.js'
+
+WIP_INFO_FORMAT_VERSION = 1
+
+class WipListener
+  # (NamespaceStorage, String | null)
+  constructor: (@storage, storageTag) ->
+    @storagePrefix = if storageTag? then "#{storageTag}:" else ""
+    @nlogoSource   = null
+    @session       = null
+    @reverted      = null
+
+  getWipKey: () ->
+    "#{@storagePrefix}#{@nlogoSource.getWipKey()}"
+
+  # (SessionLite) => Unit
+  setSession: (session) ->
+    @session = session
+    # This is necessary in the case where we didn't tell the session there was WIP at initialization becase we wanted to
+    # avoid the "loaded from cache" popup.  It's a bit icky, I know. -Jeremy B January 2023
+    if @reverted?
+      @session.widgetController.ractive.set('workInProgressState', 'enabled-with-reversion')
+
+    else
+      wipKey = @getWipKey()
+      @notifyOfWorkInProgress(@storage.hasKey(wipKey))
+
+    return
+
+  # () => String
+  getCurrentNlogo: () ->
+    @session.getNlogo()
+
+  # () => String
+  getModelTitle: () ->
+    @session.modelTitle()
+
+  # () => Unit
+  notifyOfWorkInProgress: (hasWip) ->
+    state = if hasWip then 'enabled-with-wip' else 'enabled-and-empty'
+    @session.widgetController.ractive.set('workInProgressState', state)
+    return
+
+  # (NamespaceStorage, NlogoSource, String) => WipInfo | null
+  getWip: () ->
+    wipKey   = @getWipKey()
+    maybeWip = @storage.get(wipKey)
+    if maybeWip? then maybeWip else null
+
+  # () => Unit
+  revertWip: () ->
+    wipKey = @getWipKey()
+    @reverted = @storage.get(wipKey)
+    @storage.remove(wipKey)
+    return
+
+  # () => Unit
+  undoRevert: () ->
+    wipKey = @getWipKey()
+    @storage.set(wipKey, @reverted)
+    @reverted = null
+    return
+
+  # (String, String, String) => Unit
+  _storeWipInfo: (wipKey, newNlogo, title) ->
+    wipInfo = {
+      version:   WIP_INFO_FORMAT_VERSION
+      title:     title
```

Comment:
This is another instance of being able to simply write `title` instead of `title:     title`